substrate/primitives/state-machine/src/ext.rs

```diff
@@ -733,24 +731,71 @@ impl<'a> StorageAppend<'a> {
 		Self(storage)
 	}
 
+	/// Extract current length if defined.
+	pub fn extract_nb_appends(&self) -> Option<u32> {
+		let len = u32::from(Compact::<u32>::decode(&mut &self.0[..]).ok()?);
+		Some(len)
+	}
+
+	/// Replace current length if defined.
+	pub fn replace_nb_appends(&mut self, old_length: Option<u32>, new_length: u32) {
+		let encoded_len = old_length.map(|l| Compact::<u32>::compact_len(&l)).unwrap_or(0);
+		let encoded_new = Compact::<u32>(new_length).encode();
+		if encoded_len > encoded_new.len() {
+			let diff = encoded_len - encoded_new.len();
+			*self.0 = self.0.split_off(diff);
+		} else if encoded_len < encoded_new.len() {
+			let diff = encoded_new.len() - encoded_len;
+			// Non constant change
+			for _ in 0..diff {
+				self.0.insert(0, 0);
+			}
+		}
+		self.0[0..encoded_new.len()].copy_from_slice(&encoded_new);
```

Comment:
This annoyance is only needed since we use `Compact<u32>` instead of `u32`?

Reply:
yes, this is equivalent to code from EncodeAppend of scale lib.
pretty nasty copy: fix size or compact size could have been stored at end (with reverse order of byte) would be better.
halas cannot change it due to it being part of a host call https://github.com/paritytech/polkadot-sdk/blob/a42a47f81d91ccf2f885ff01749185cd4c793278/substrate/primitives/io/src/lib.rs#L298